test/cpp/jit/test_concat_opt.cpp

```diff
@@ -0,0 +1,451 @@
+              %4: Float(32, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu)):
+          %5 : int = prim::Constant[value=0]()
+          %features.1 : Tensor[] = prim::ListConstruct(%0)
+          %concat.1 : Float(64, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::cat(%features.1, %5)
+
+          %features.2 : Tensor[] = prim::ListConstruct(%0, %1)
+          %concat.2 : Float(96, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::cat(%features.2, %5)
+
+          %features.3 : Tensor[] = prim::ListConstruct(%0, %1, %2)
+          %concat.3 : Float(128, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::cat(%features.3, %5)
+
+          %features.4 : Tensor[] = prim::ListConstruct(%0, %1, %2, %3)
+          %concat.4 : Float(160, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::cat(%features.4, %5)
+
+          %features.5 : Tensor[] = prim::ListConstruct(%0, %1, %2, %3, %4)
+          %concat.5 : Float(192, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::cat(%features.5, %5)
+
+          %res : Tensor[] = prim::ListConstruct(%concat.1, %concat.2, %concat.3, %concat.4, %concat.5)
+          return (%res)
+      )IR";
+  parseIR(input, graph.get());
+  std::vector<at::Tensor> inputs = {
+      at::rand({64, 56, 56}, at::kCPU),
+      at::rand({32, 56, 56}, at::kCPU),
+      at::rand({32, 56, 56}, at::kCPU),
+      at::rand({32, 56, 56}, at::kCPU),
+      at::rand({32, 56, 56}, at::kCPU)};
+  auto orig_outputs = runGraph(graph, inputs);
+
+  OptimizeConcat(graph);
+  graph->lint();
+  auto opt_outputs = runGraph(graph, inputs);
+
+  checkOutputs(orig_outputs, opt_outputs);
+
+  testing::FileCheck()
+      .check_count("= aten::cat(", 0, /*exactly*/ true)
+      ->check_count("= aten::empty(", 1, /*exactly*/ true)
+      ->check_count("= aten::copy_(", 5, /*exactly*/ true)
+      ->check_count("= aten::cat(", 0, /*exactly*/ true)
+      ->check_count("= aten::empty(", 0, /*exactly*/ true)
+      ->run(*graph);
+}
+
+TEST(OptimizeConcatTest, ConcatWithoutResultShape) {
+  auto graph = std::make_shared<Graph>();
+
+  const std::string input =
+      R"IR(
+        graph(%0: Float(64, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu),
+              %1: Float(32, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu)):
+          %2 : int = prim::Constant[value=0]()
+          %3 : float = prim::Constant[value=0.5]()
+          # CHECK: clamp_max
+          %4 : Float(64, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::clamp_max(%0, %3)
+          # CHECK: clamp_max
+          %5 : Float(32, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::clamp_max(%1, %3)
+          # CHECK: prim::ListConstruct
+          %6 : Tensor[] = prim::ListConstruct(%4, %5)
+          # CHECK: aten::cat
+          %7 : Tensor = aten::cat(%6, %2)
```

Comment:
Did not understand this example:
- Why do we need to know the shape of %7 for the concat optimization?
- Couldn't JIT automatically figure out the shape of %7 based on that of %6 and %2?

Reply:
We can figure that out but that would be doing shape inference here. I'd prefer to keep that a separate pass and just handle `cat` optimizations here.

Reply:
Right. It makes sense to separate concerns on inferring shapes to a different pass.